### DISCUSSION TITLE
Add block with component but styles are disappear

### DISCUSSION THREAD

**Discussion by momu-2016** [GrapesJS/grapesjs#3872] [Q&A] [answered]
at first i save my component as html and css to mysql,and i load it from block ,just like:

` this.em.get("BlockManager").add(model.get("id") + index, {
              label: `<img src="${img}"/> `,
              content: {
                 content:endhtml,
                serverid:item.id,
              },
              category: model.get("id"),
              type: 1,
            });`
the endhtml contain html js and css.

it shows well.but i find all the Component related custom trait all lost, so i change to save component and styles as json string to mysql,and  i try to load block like this:
` this.em.get("BlockManager").add(model.get("id") + index, {
              label: `<img src="${img}"/> `,
              content: {
                components:JSON.parse(item.component),
                styles:JSON.parse(item.component_style),
                serverid:item.id,
              },
              category: model.get("id"),
              type: 1,
            });
          });`

the component's trait is ok,but style disappear.

need to help
thanks all!

**Comment by artf** [comment]
What is the content of `JSON.parse(item.component_style)`?

**Reply by momu-2016** [reply]
![image](https://user-images.githubusercontent.com/18502726/138197728-6bed6bbe-5dc2-4f8b-92e4-6936101eae7f.png)
gis-styles about the component that i saved  is the content of component_style.

and I have found a solution.
first i add the component to the block's  content,and add a custom attribute component_style 
![image](https://user-images.githubusercontent.com/18502726/138198238-1a6850ee-a222-497a-a600-0fa0a7b71146.png)

the editor will render the component without style when dropped

and i create a listener for events
![image](https://user-images.githubusercontent.com/18502726/138198415-b449bf75-81cd-469e-93ca-f8237a537ca0.png)
if the block has the attribute of component_style ,then set the component style to editor.

although I know it looks stupid, but it does solve my current needs.

Anyway, thank you for your reply and such an amazing project.

**Comment by artf** [comment, marked as answer]
@momu-2016 Please check the guide about [Component & CSS](https://grapesjs.com/docs/modules/Components.html#components-css) and also what [you shouldn't be doing](https://grapesjs.com/docs/modules/Blocks.html#important-caveats) in your Blocks (now you're doing the opposite).
Your `styles:JSON.parse(item.component_style)` concept should work properly just move them in a custom component and ensure to have the latest version as it was added recently.

So, I highly recommend creating custom components (similar to what you have done with blocks) and then adding the custom block by simply using `content: { type: 'your-custom-component-id' }`
